input-events/input-events-typing.html

```diff
@@ -1,82 +1,311 @@
+(function() {
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        const message = 'Hello';
+        await test_driver.send_keys(document.getElementById('rich'), message);
+        // 10 events (5 beforeinput + 5 input events)
+        assert_equals(inputEventsLog.length, 10);
+        for (let i = 0; i < inputEventsLog.length; i += 2) {
+            const beforeInputEvent = inputEventsLog[i];
+            const inputEvent = inputEventsLog[i + 1];
+            assert_equals(beforeInputEvent.type, 'beforeinput');
+            assert_equals(inputEvent.type, 'input');
+            assert_equals(beforeInputEvent.inputType, 'insertText');
+            assert_equals(inputEvent.inputType, 'insertText');
+            assert_equals(beforeInputEvent.data, inputEvent.data);
+            assert_equals(inputEvent.data, message[i / 2]);
+            assert_equals(beforeInputEvent.state + message[i / 2], inputEvent.state);
+        }
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on text typing');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await test_driver.send_keys(document.getElementById('rich'), "\uE006"); // Return
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'insertParagraph');
+        assert_equals(inputEvent.inputType, 'insertParagraph');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing RETURN');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await new test_driver.Actions()
+            .keyDown('\uE008') // Shift
+            .keyDown('\uE006') // Enter
+            .keyUp('\uE006')
+            .keyUp('\uE008')
+            .send();
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'insertLineBreak');
+        assert_equals(inputEvent.inputType, 'insertLineBreak');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing Shift+RETURN');
+
+    promise_test(async () => {
+        resetEditor('<p>Preexisting <i id="caret">c</i>ontent</p>');
+        await test_driver.click(document.querySelector('#caret'));
```

Comment:
nit... do the following, before the await? Less lookups. 

```
const caret = document.getElementById('caret');
```

Reply:
done